lib/charms/opensearch/v0/helper_cluster.py

```diff
@@ -77,106 +75,39 @@ def get_cluster_settings(
-        # to reset the roles to their expected default values so that "roles re-balancing"
-        # logic (node_with_new_roles) can be safely applied: only change the last node.
-        # Nothing changes to the conf of the nodes if the roles were previously auto-generated.
-        valid_nodes = ClusterTopology.refill_node_with_default_roles(app_name, nodes)
-
-        nodes_by_name = dict([(node.name, node) for node in nodes])
-
-        # compute node with new roles - only returns 1 changed node
-        updated_node = ClusterTopology.node_with_new_roles(app_name, valid_nodes)
-        if updated_node:
-            nodes_by_name[updated_node.name] = updated_node
-
-        return nodes_by_name
-
-    @staticmethod
-    def refill_node_with_default_roles(app_name: str, nodes: List[Node]) -> List[Node]:
-        """Refill the roles of a list of nodes with default values for re-computing.
-
-        This method works hand in hand with node_with_new_roles which assumes a clean
-        base (regarding the auto-generation logic) and only applies changes to 1 node.
-        """
+        if not nodes:
+            return {}
+        logger.debug(f"Roles before re-balancing {({node.name: node.roles for node in nodes})=}")
+        nodes_by_name = {}
+        current_cluster_nodes = []
+        for node in nodes:
+            if node.app_name == app_name:
+                current_cluster_nodes.append(node)
+            else:
+                # Leave node unchanged
+                nodes_by_name[node.name] = node
         base_roles = ["data", "ingest", "ml", "coordinating_only"]
         full_roles = base_roles + ["cluster_manager"]
-
-        current_cluster_nodes = [node for node in nodes if node.app_name == app_name]
-        current_cm_eligible = [node for node in current_cluster_nodes if node.is_cm_eligible()]
-
-        # we check if the roles were previously balanced in accordance with the auto-generation
-        # logic, in which the max difference between CMs and Non-CMs is 1 node (to keep quorum)
-        unbalanced = len(current_cm_eligible) < len(current_cluster_nodes) - 1
-
-        updated_nodes = []
+        highest_unit_number = max(node.unit_number for node in current_cluster_nodes)
         for node in current_cluster_nodes:
             # we do this in order to remove any non-default role / add any missing default role
-            new_roles = full_roles if unbalanced or node.is_cm_eligible() else base_roles
-            updated = Node(
+            if len(current_cluster_nodes) % 2 == 0 and node.unit_number == highest_unit_number:
+                roles = base_roles
+            else:
+                roles = full_roles
+
+            nodes_by_name[node.name] = Node(
                 name=node.name,
-                roles=new_roles,
+                roles=roles,
                 ip=node.ip,
                 app_name=node.app_name,
+                unit_number=node.unit_number,
```

Comment:
nit: Given the unit_number can be easily be inferred from `Node.name` as follows: `int(self.name.split("-")[-1])` it would make sense to not have it as part of of the constructor arguments but rather as a `@validator` on the Node model class itself